jeeps/gpsserial.cc

```diff
@@ -23,671 +23,298 @@
+   * Certified driver", will crash the OS on a write of zero bytes.
+   * We get such writes from upstream when there are zero payload
+   * bytes.  SO we trap those here to stop Keyspan & Windows from
+   * nuking the system.
+   */
+  if (size == 0) {
+    return 0;
+  }
 
+  qint64 len = h->sp.write(static_cast<const char*>(obuf), size);
 
-/* @func GPS_Serial_Off ***********************************************
-**
-** Done with port
-**
-** @param [r] port [const char *] port
-** @param [r] fd [int32_t ] file descriptor
-**
-** @return [int32_t] success
-************************************************************************/
+  if (len != size) {
+    fatal("Write error.  Wrote %d of %d bytes.\n", static_cast<int>(len), size);
+  }
 
-int32_t GPS_Serial_Off(gpsdevh* dh)
+  /* Call waitForBytesWritten to process IO in blocking mode.
+   */
+  bool ok = h->sp.waitForBytesWritten(10);
+  if (!ok) {
+    h->sp.clearError();
+  }
+  return len;
+}
+
+int32_t GPS_Serial_Read(gpsdevh* dh, void* ibuf, int size)
 {
+  auto* h = reinterpret_cast<serial_data_handle*>(dh);
 
-  if (!GPS_Serial_Close(dh)) {
-    GPS_Error("Error Closing port");
-    gps_errno = HARDWARE_ERROR;
-    return 0;
-  }
-  xfree(dh);
-  return 1;
+  /* Process IO */
+  (void) GPS_Serial_Chars_Ready(dh);
+
+  qint64 cnt = h->sp.read(static_cast<char*>(ibuf), size);
+  return cnt;
 }
 
 // Based on information by Kolesár András from
 // http://www.manualslib.com/manual/413938/Garmin-Gps-18x.html?page=32
-int32_t GPS_Serial_Set_Baud_Rate(gpsdevh* fd, int br)
+int32_t GPS_Serial_Set_Baud_Rate(gpsdevh* dh, int br)
 {
-
-  struct termios tty;
   static UC data[4];
   GPS_Packet tra;
   GPS_Packet rec;
-
-  speed_t speed = mkspeed(br);
+  auto* h = reinterpret_cast<serial_data_handle*>(dh);
 
   // Turn off all requests by transmitting packet
   GPS_Util_Put_Short(data, 0);
   GPS_Make_Packet(&tra, 0x1c, data, 2);
-  if (!GPS_Write_Packet(fd,tra)) {
+  if (!GPS_Write_Packet(dh,tra)) {
     return gps_errno;
   }
-  if (!GPS_Get_Ack(fd, &tra, &rec)) {
+  if (!GPS_Get_Ack(dh, &tra, &rec)) {
     return gps_errno;
   }
 
   GPS_Util_Put_Int(data, br);
   GPS_Make_Packet(&tra, 0x30, data, 4);
-  if (!GPS_Write_Packet(fd,tra)) {
+  if (!GPS_Write_Packet(dh,tra)) {
     return gps_errno;
   }
-  if (!GPS_Get_Ack(fd, &tra, &rec)) {
+  if (!GPS_Get_Ack(dh, &tra, &rec)) {
     return gps_errno;
   }
 
   // Receive IOP_BAUD_ACPT_DATA
-  if (!GPS_Packet_Read(fd, &rec)) {
+  if (!GPS_Packet_Read(dh, &rec)) {
     return gps_errno;
   }
 
   // Acknowledge new speed
-  if (!GPS_Send_Ack(fd, &tra, &rec)) {
+  if (!GPS_Send_Ack(dh, &tra, &rec)) {
     return gps_errno;
   }
-  GPS_Device_Flush(fd);
-  GPS_Device_Wait(fd);
+  GPS_Device_Flush(dh);
+  GPS_Device_Wait(dh);
 
   // Sleep for a small amount of time, about 100 milliseconds,
   // to make sure the packet was successfully transmitted to the GPS unit.
-    QThread::usleep(100000);
+  QThread::usleep(100000);
```

Comment:
The original code looks wrong to me!
GPS_Device_Flush on unix calls tcflush(), which I think I correclty translated to QSerialPort::clear().
But tcflush discards data, it doesn't send it on its way.
```
tcflush() discards data written to the object referred to by fd but not transmitted, or data received but not read, depending on the value of queue_selector:
```
What [GPS_18x_Tech_Specs.pdf](https://static.garmincdn.com/pumac/GPS_18x_Tech_Specs.pdf) says is to wait a small amount of time to make sure the previous packet (the ACK of IOP_BAUD_ACPT_DATA) is complete.  So it sounds to me like we wanted to call tcdrain() instead of tcflush(), which would translate to QSerialPort::flush().  QSerialPort::flush() may not be necessary after QSerialPort::waitForBytesWritten.

I note that GPS_18x_Tech_Specs were updated 1/2024!

I cant test this as the GPS12 doesn't support baud rate changes.